### PR TITLE
47041437_Improved 3D U-Net segmentation on 3D prostate MRI scans

### PR DESCRIPTION
**Pull request by Watermelon132**
# Project Overview

**Name:** Sara Alaei  
**Student Number:** 47041437  
**Task:** Improved 3D U-Net Segmentation on 3D Prostate MRI Scans  
**Difficulty:** Hard

## Summary

Both the original and improved U-Net model architectures were investigated. The DSC plots and image segmentation outputs are included in the README.

## Source Files

- `dataset.py` – Loads dataset
- `module_unet3D.py` – Original 3D U-Net model
- `module_improvedunet3D.py` – Improved 3D U-Net model
- `train.py` – `main()` training loop
- `predict.py` – Loads saved model to create 2D axial image slices of original and segmented sample

## Results

Dice similarity coefficient of **0.75** and **0.8** were achieved for the 3D U-Net and improved 3D U-Net models, respectively.

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**

Recognition problem: total 14.5 Marks

Problem Solving: (5)
•	Algorithm solves the problem and achieves a Dice score of above 0.7 for all labels. 
Algorithm implementation: (3)
•	Algorithm implemented in PyTorch for the UNet3D model and functions.
Design: Good coding design (1)
Commenting: missing comments for the two module files. (0.5)
Difficulty: (5)
•	From document, the improved UNet3D was mentioned but in the code (module_improvedunet3D.py), the 3D UNet was built instead of the improved UNet which includes deep supervision layers.
•	3D UNet – Normal

**Comment by Watermelon132** [comment]
Hi, 
I am not sure what you are referring to here, I worked on both problems so I included the results from both, I included the results for both the normal and improved version. 

Also, the compiled version includes the improved the UNET 3d. **If you scroll to the imports** you will see that I have commented out the normal 3D. 

Could you please let me know if this changes my score?
Thanks.

**Comment by Watermelon132** [comment]
Hi,

Thanks for your feedback on my improved UNet 3D model. I just wanted to point out that the version I submitted includes some improvements over the original UNet 3D, like adding batch normalization layers and using trilinear upsampling for smoother results. It also uses inplace ReLU to optimize memory usage. While I know it’s missing the deep supervisory layers, I focused a lot on making sure the model worked with my data and spent a significant amount of time fine-tuning it.

Given the effort I put into adapting the model, I feel a bit discouraged by the grade. Getting just a 5 without partial credit seems a bit harsh, so I’d really appreciate it if you could reconsider. I’m also open to any suggestions on how I can improve the model further - please let me know how much time I have if improvements I'm allowed to make adjustments. 

Thanks for your time!

Regards,
Sara

**Comment by celinsze** [comment]
Hi Sara, marks are not finalised yet and that was the initial marking and comments which would be considered for the final marks.

**Comment by Watermelon132** [comment]
Hi,

I have now switched to a more robust version of the improved UNet 3D model that incorporates the expected deep supervision layers and achieves a **Dice Similarity Coefficient (DSC) of at least 0.8 across all labels**.

Since I can still make changes (as mentioned on EdStem), I have made the following updates based on your feedback:

**1. Updated Model**: Switched to an improved 3D U-Net architecture in ```module_improvedunet3D.py```, which **includes deep supervision layers** and other key improvements.

**2. Updated Documentation:** Updated README.md with new loss plots, image slices, and references to reflect the results of training with the new model. 

**3. Added Comments:** Added docstrings and comments in both module files to make the code easier to follow.

Hopefully, this version meets the criteria for an improved 3D U-Net model.

Regards,
Sara

**Comment by shakes76** [comment]
Feedback done +2, Segmentations not one hot encoded -2 additionally
